### PR TITLE
Add cloud function/job calls and user password reset.

### LINKED ISSUE #44: iOS 14 User logout does not erase session
Hi All - Wanted to check if I am having an implementation issue.  I have tried calling logout both synchronously and asynchronously.  No trouble removing the keychain object, but the session object in database persists.

Anybody having this issue in iOS 14?

### PR DESCRIPTION
**Pull request by cbaker6**
- [x] Also adds more command tests for User and cleans up user docs
- [x] Add server options to ParseUsers that were missing
- [x] Ensures `ParseEncoder` is used for Parse types while json encoder is used for others
- [x] Playground examples 
- [x] Batches send in waves of 50 objects. This can be changed by the developer by changing `batchLimit` when calling `saveAll` or `deleteAll` 
- [x] Commented out some of the Batch code for saving child `ParseObjects` that wasn't being used because it's in a non-working state. It's not needed for now, child objects are sent to the server 1 by 1 until a fix. This may cause more http overhead if there's a lot of unsaved children. I don't see this happening much unless a developer attempts to saveAll "many" `ParseObjects` that have "many" unsaved children. In this case, `URLSession` and the parse-server should still be able to handle the load. Batching still works for `ParseObjects`, just not for unsaved children.
- [x] Bug fix #44 where the SDK was calling the wrong endpoint to log out of users.
- [x] User and Installation now talk to their respective endpoints directly instead of through `/classes`
- [x] Adds verificationEmailRequest
- [x] Improves server response error handling
- [x] Add Swift 5.2 build to CI for backwards compatibility test

### CODE REVIEW

**Comment by cbaker6** [comment]
@TomWFox can you review the docs?

@pranjalsatija can you review?

File: Sources/ParseSwift/Objects/ParseUser.swift [hunk 1]
```diff
@@ -175,20 +198,136 @@ extension ParseUser {
...
        }
     }
 }
 
+// MARK: Password Reset
+extension ParseUser {
+
+    /**
+     Requests *synchronously* a password reset email to be sent to the specified email address
+     associated with the user account. This email allows the user to securely reset their password on the Parse site.
```
Comment (TomWFox) [inline Sources/ParseSwift/Objects/ParseUser.swift:242]: Same as my other comment.

File: Sources/ParseSwift/Objects/ParseUser.swift [hunk 2]
```diff
@@ -175,20 +198,136 @@ extension ParseUser {
...
+    */
+    public static func passwordReset(email: String, options: API.Options = []) throws {
+        if let error = try passwordResetCommand(email: email).execute(options: options) {
+            throw error
+        }
+    }
+
+    /**
+     Requests *asynchronously* a password reset email to be sent to the specified email address
+     associated with the user account. This email allows the user to securely reset their password on the Parse site.
```
Comment (TomWFox) [inline Sources/ParseSwift/Objects/ParseUser.swift:254]: and here.

File: Sources/ParseSwift/Types/ParseFile.swift [hunk 3]
```diff
@@ -43,7 +43,7 @@ public struct ParseFile: Fileable, Savable, Fetchable, Deletable, Hashable {
     public var localURL: URL?
 
     /**
-     The link to the file online that should be downloaded.
+     The link to the file online that should be fetched before uploading to the parse server.
```
Comment (TomWFox) [inline Sources/ParseSwift/Types/ParseFile.swift:46]:
```suggestion
     The link to the file online that should be fetched before uploading to the Parse Server.
```

File: Sources/ParseSwift/Types/ParseFile.swift [hunk 4]
```diff
@@ -97,8 +100,11 @@ public struct ParseFile: Fileable, Savable, Fetchable, Deletable, Hashable {
      - parameter mimeType: Specify the Content-Type header to use for the file,  for example
      "application/pdf". The default is nil. If no value is specified the file type will be inferred from the file
      extention of `name`.
-     - parameter metadata: Optional key value pairs to be stored with file object
+     - parameter metadata: Optional key value pairs to be stored with file object.
      - parameter tags: Optional key value pairs to be stored with file object
+     - note: `metadata` and `tags` is file adapter specific and not supported by all file adapters.
+     For more, see details on the
+     [S3 adapter](https://github.com/parse-community/parse-server-s3-adapter#adding-metadata-and-tags)
```
Comment (TomWFox) [inline Sources/ParseSwift/Types/ParseFile.swift:107]:
```suggestion
     - parameter tags: Optional key value pairs to be stored with file object.
     - note: `metadata` and `tags` is file adapter specific and not supported by all file adapters.
     For more, see details on the
     [S3 adapter](https://github.com/parse-community/parse-server-s3-adapter#adding-metadata-and-tags).
```

File: Sources/ParseSwift/Types/ParseFile.swift [hunk 5]
```diff
@@ -122,6 +128,9 @@ public struct ParseFile: Fileable, Savable, Fetchable, Deletable, Hashable {
      extention of `name`.
      - parameter metadata: Optional key value pairs to be stored with file object
      - parameter tags: Optional key value pairs to be stored with file object
+     - note: `metadata` and `tags` is file adapter specific and not supported by all file adapters.
+     For more, see details on the
+     [S3 adapter](https://github.com/parse-community/parse-server-s3-adapter#adding-metadata-and-tags)
```
Comment (TomWFox) [inline Sources/ParseSwift/Types/ParseFile.swift:133]:
```suggestion
     - parameter metadata: Optional key value pairs to be stored with file object.
     - parameter tags: Optional key value pairs to be stored with file object.
     - note: `metadata` and `tags` is file adapter specific and not supported by all file adapters.
     For more, see details on the
     [S3 adapter](https://github.com/parse-community/parse-server-s3-adapter#adding-metadata-and-tags).
```

**Review by TomWFox — CHANGES_REQUESTED** [review]
Just a few minor changes... otherwise LGTM

**Comment by cbaker6** [comment]
@TomWFox can you add `spm-test-5_2` to the branch protections? This ensures we are still Swift backwards compatible, which I've managed to fix/break in the last couple of PRs. The issue is Swift 5.3 reduced the need for `self.` inside of closures, so the compiler for 5.3 doesn't complain when you don't use it. Older versions of Swift require the use of `self` inside of closures.

**Review by TomWFox — APPROVED** [review]

**Comment by cbaker6** [comment]
@TomWFox can look at the documentation I added for the `batchLimit` argument of `saveAll/deleteAll` in `ParseObject.swift`, `ParseUser.swift`, `ParseInstallation.swift`? There are two of these in each file.

File: Sources/ParseSwift/Objects/ParseObject.swift [hunk 6]
```diff
@@ -54,42 +59,166 @@ public extension Sequence where Element: ParseObject {
 
     /**
      Saves a collection of objects *synchronously* all at once and throws an error if necessary.
-
-     - parameter options: A set of options used to save objects. Defaults to an empty set.
+     - parameter batchLimit: The amount of objects to send in each batch. If the items to be batched
```
Comment (TomWFox) [inline Sources/ParseSwift/Objects/ParseObject.swift:62]: I think this is slightly better...
```suggestion
     - parameter batchLimit: The maximum number of objects to send in each batch. If the items to be batched
```

File: Sources/ParseSwift/Objects/ParseObject.swift [hunk 7]
```diff
@@ -54,42 +59,166 @@ public extension Sequence where Element: ParseObject {
...
+            returnBatch.append(contentsOf: currentBatch)
+        }
+        return returnBatch
     }
 
     /**
      Saves a collection of objects all at once *asynchronously* and executes the completion block when done.
-
-     - parameter options: A set of options used to save objects. Defaults to an empty set.
+     - parameter batchLimit: The amount of objects to send in each batch. If the items to be batched
```
Comment (TomWFox) [inline Sources/ParseSwift/Objects/ParseObject.swift:132]:
```suggestion
     - parameter batchLimit: The maximum number of objects to send in each batch. If the items to be batched
```

File: Sources/ParseSwift/Objects/ParseObject.swift [hunk 8]
```diff
@@ -167,8 +296,10 @@ public extension Sequence where Element: ParseObject {
 
     /**
      Deletes a collection of objects *synchronously* all at once and throws an error if necessary.
-
-     - parameter options: A set of options used to delete objects. Defaults to an empty set.
+     - parameter batchLimit: The amount of objects to send in each batch. If the items to be batched
```
Comment (TomWFox) [inline Sources/ParseSwift/Objects/ParseObject.swift:299]:
```suggestion
     - parameter batchLimit: The maximum number of objects to send in each batch. If the items to be batched
```

File: Sources/ParseSwift/Objects/ParseObject.swift [hunk 9]
```diff
@@ -180,17 +311,27 @@ public extension Sequence where Element: ParseObject {
...
+            returnBatch.append(contentsOf: currentBatch)
+        }
+        return returnBatch
     }
 
     /**
      Deletes a collection of objects all at once *asynchronously* and executes the completion block when done.
-
-     - parameter options: A set of options used to delete objects. Defaults to an empty set.
+     - parameter batchLimit: The amount of objects to send in each batch. If the items to be batched
```
Comment (TomWFox) [inline Sources/ParseSwift/Objects/ParseObject.swift:331]:
```suggestion
     - parameter batchLimit: The maximum number of objects to send in each batch. If the items to be batched
```

File: Sources/ParseSwift/Objects/ParseUser.swift [hunk 10]
```diff
@@ -487,62 +662,189 @@ extension ParseUser {
...
 }
 
 // MARK: Batch Support
 public extension Sequence where Element: ParseUser {
 
     /**
      Saves a collection of users *synchronously* all at once and throws an error if necessary.
-
-     - parameter options: A set of options used to save users. Defaults to an empty set.
+     - parameter batchLimit: The amount of objects to send in each batch. If the items to be batched
```
Comment (TomWFox) [inline Sources/ParseSwift/Objects/ParseUser.swift:685]:
```suggestion
     - parameter batchLimit: The maximum number of objects to send in each batch. If the items to be batched
```

File: Sources/ParseSwift/Objects/ParseUser.swift [hunk 11]
```diff
@@ -487,62 +662,189 @@ extension ParseUser {
...
+        }
+        try? Self.Element.updateKeychainIfNeeded(returnBatch.compactMap {try? $0.get()})
+        return returnBatch
     }
 
     /**
      Saves a collection of users all at once *asynchronously* and executes the completion block when done.
-
-     - parameter options: A set of options used to save users. Defaults to an empty set.
+     - parameter batchLimit: The amount of objects to send in each batch. If the items to be batched
```
Comment (TomWFox) [inline Sources/ParseSwift/Objects/ParseUser.swift:756]:
```suggestion
     - parameter batchLimit: The maximum number of objects to send in each batch. If the items to be batched
```

File: Sources/ParseSwift/Objects/ParseUser.swift [hunk 12]
```diff
@@ -624,8 +926,10 @@ public extension Sequence where Element: ParseUser {
 
     /**
      Deletes a collection of users *synchronously* all at once and throws an error if necessary.
-
-     - parameter options: A set of options used to delete users. Defaults to an empty set.
+     - parameter batchLimit: The amount of objects to send in each batch. If the items to be batched
```
Comment (TomWFox) [inline Sources/ParseSwift/Objects/ParseUser.swift:929]:
```suggestion
     - parameter batchLimit: The maximum number of objects to send in each batch. If the items to be batched
```

File: Sources/ParseSwift/Objects/ParseUser.swift [hunk 13]
```diff
@@ -638,20 +942,28 @@ public extension Sequence where Element: ParseUser {
...
         try? Self.Element.updateKeychainIfNeeded(compactMap {$0})
-        return returnResults
+        return returnBatch
     }
 
     /**
      Deletes a collection of users all at once *asynchronously* and executes the completion block when done.
-
-     - parameter options: A set of options used to delete users. Defaults to an empty set.
+     - parameter batchLimit: The amount of objects to send in each batch. If the items to be batched
```
Comment (TomWFox) [inline Sources/ParseSwift/Objects/ParseUser.swift:963]:
```suggestion
     - parameter batchLimit: The maximum number of objects to send in each batch. If the items to be batched
```

File: Sources/ParseSwift/Objects/ParseInstallation.swift [hunk 14]
```diff
@@ -470,62 +520,190 @@ extension ParseInstallation {
...
 }
 
 // MARK: Batch Support
 public extension Sequence where Element: ParseInstallation {
 
     /**
      Saves a collection of installations *synchronously* all at once and throws an error if necessary.
-
-     - parameter options: A set of options used to save installations. Defaults to an empty set.
+     - parameter batchLimit: The amount of objects to send in each batch. If the items to be batched
```
Comment (TomWFox) [inline Sources/ParseSwift/Objects/ParseInstallation.swift:543]:
```suggestion
     - parameter batchLimit: The maximum number of objects to send in each batch. If the items to be batched
```

File: Sources/ParseSwift/Objects/ParseInstallation.swift [hunk 15]
```diff
@@ -470,62 +520,190 @@ extension ParseInstallation {
...
+        }
+        try? Self.Element.updateKeychainIfNeeded(returnBatch.compactMap {try? $0.get()})
+        return returnBatch
     }
 
     /**
      Saves a collection of installations all at once *asynchronously* and executes the completion block when done.
-
-     - parameter options: A set of options used to save installations. Defaults to an empty set.
+     - parameter batchLimit: The amount of objects to send in each batch. If the items to be batched
```
Comment (TomWFox) [inline Sources/ParseSwift/Objects/ParseInstallation.swift:615]:
```suggestion
     - parameter batchLimit: The maximum number of objects to send in each batch. If the items to be batched
```

File: Sources/ParseSwift/Objects/ParseInstallation.swift [hunk 16]
```diff
@@ -607,8 +785,10 @@ public extension Sequence where Element: ParseInstallation {
 
     /**
      Deletes a collection of installations *synchronously* all at once and throws an error if necessary.
-
-     - parameter options: A set of options used to delete installations. Defaults to an empty set.
+     - parameter batchLimit: The amount of objects to send in each batch. If the items to be batched
```
Comment (TomWFox) [inline Sources/ParseSwift/Objects/ParseInstallation.swift:788]:
```suggestion
     - parameter batchLimit: The maximum number of objects to send in each batch. If the items to be batched
```

File: Sources/ParseSwift/Objects/ParseInstallation.swift [hunk 17]
```diff
@@ -621,20 +801,29 @@ public extension Sequence where Element: ParseInstallation {
...
         try? Self.Element.updateKeychainIfNeeded(compactMap {$0})
-        return returnResults
+        return returnBatch
     }
 
     /**
      Deletes a collection of installations all at once *asynchronously* and executes the completion block when done.
-
-     - parameter options: A set of options used to delete installations. Defaults to an empty set.
+     - parameter batchLimit: The amount of objects to send in each batch. If the items to be batched
```
Comment (TomWFox) [inline Sources/ParseSwift/Objects/ParseInstallation.swift:823]:
```suggestion
     - parameter batchLimit: The maximum number of objects to send in each batch. If the items to be batched
```

**Review by TomWFox — CHANGES_REQUESTED** [review]

File: Sources/ParseSwift/Objects/ParseObject.swift (hunk 6 above)
Comment (cbaker6) [reply Sources/ParseSwift/Objects/ParseObject.swift:62]: Agreed

**Review by TomWFox — APPROVED** [review]

**Pull request merged**